src/twisted/test/test_iosim.py

```diff
@@ -39,3 +44,85 @@ def test_writeSequence(self):
         a.writeSequence([b"b", b"c", b"d"])
 
         self.assertEqual(b"".join(a.stream), b"abcd")
+
+
+
+@implementer(IPushProducer)
+class StrictPushProducer(object):
```

Comment:
Probably.  Automat won't let you inspect the current state of the machine though, which makes it a little annoying.  I'd basically have to continue to maintain the same manual state as the code maintains now, plus have the automat stuff.  Probably not a net win.

(@glyph)